src/django_pg_migration_tools/operations.py

```diff
@@ -525,3 +583,256 @@ def describe(self) -> str:
             f"using a UNIQUE index. NOTE: Using the django_pg_migration_tools "
             f"SaferRemoveIndexConcurrently operation."
         )
+
+
+class NullsManager(base_operations.Operation):
+    def set_not_null(
+        self,
+        app_label: str,
+        schema_editor: base_schema.BaseDatabaseSchemaEditor,
+        model: type[models.Model],
+        column_name: str,
+    ) -> None:
+        """
+        The following routine takes into consideration:
```

Comment:
What does this method do, and how?

The `if/elif` conditions in the code below (and the subseries of operations within each) are hard to understand without knowing what this method does, and the series of steps it takes to achieve that.

The above information is in the commit comment. It should be in this docstring instead. See [Wints' blogpost](https://codeinthehole.com/tips/explaining/#this-applies-everywhere).

Reply:
I'm concerned about adding how/what description of this routine in the
docstrings, I prefer context over how/what. Docstrings become out-of-date as a
function grows, inducing confusion and errors.

As of now the information is available in 3 places: commit message, docs, and
tests. I'd like to retain the information in the commit message because it
reflects the information at the time the change was made, and git blame is
very powerful for that.

Reply:
I think your argument is valid when applied to internal Kraken Tech repos, where the users of the code are all developers of the code, and will be accustomed to checking git history and commit messages.

But this is a third-party utility library. The users of this code will most likely have installed it via `pip install`; they won't clone the repo, and they won't look at the tests. They may view this method by following code definitions in their IDE. The docstring as it stands, without the context of even a single line statement of what it does, and without a reference to where to get more information, is unhelpful.

> Docstrings become out-of-date as a function grows, inducing confusion and errors.

There are ~18 lines of docstring below the one line I highlighted above, all of which are subject to your concern about becoming out of date.

I don't know if there's much overlap between a docstring you'd be happy with and one that would keep me happy also. :(

But as a thought experiment, imagine you'd released the code without the fix in https://github.com/kraken-tech/django-pg-migration-tools/commit/8846dd2954b43cb86a27dfeaf5bbb8c620e8184a. Then some user of this package encountered an error that left behind the constraint when applying the migration, which failed initially, and found that the constraint was still there after the migration applied successfully. They then went digging into the code to find what happened. What docstring would help them the most to find the issue?

Reply:
> There are ~18 lines of docstring below the one line I highlighted above, all of which are subject to your concern about becoming out of date.

Yeah, that's a logical flaw in my argument. Just having any docstring is enough
for it to become out of date. A small nuance though, is that by only having
context in the docstring (that re-entrancy and idempotency is expected), the
person who found the bug would know that the code didn't follow the expected
outcome. I assume that's how you come to that conclusion that it was indeed a
problem (as you are the real bug-finder yourself in that scenario)?

I'd be up for giving more context about expected outcome − for example that
queries are expected to not lock reads/writes (at least not for very long) −
but I don't think we should explain the SQL as detailed as the docs have. I'm
not sure if that's what you are suggesting?

I agree that users will inspect code, docs, commands, etc., in different ways.
But I don't think that replicating documentation to all different ways a user
may interact with the repo is the solution

Reply:
Thanks. I'm more relieved now that it's clear this isn't about detailing out
information about the SQL queries, what they exactly do, and further context.

With a few small alterations for clarification here is the change: cbe227bcc8766f39404727b9a3460835b357a387.

Reply:
Thanks, those clarifications LGTM.